README.md
# Elementary Cellular Automata

Terminal visualization of elementary cellular automata.
An elementary cellular automaton is visualized as filled/empty cells on a two-dimensional grid that evolves from top to bottom.  A elementary cellular automata has a single rule that determines how it evolves.
Each cell in an elementary cellular automata can only have a value of 'on' or 'off'.
The value of a cell in the current generation is determined by the state of the three cells above it in the previous generation.
The top row (or first generation) is the initial state, and subsequent generations are calculated based on the value of cells in the current generation based on a rule.

### Rules
A rule is a number between 0 - 255 that corresponds to an 8-bit value that represents how three adjacent cells determine the value of the middle cell in the next generation.

Rules:
```
_______  _______  _______  _______  _______  _______  _______  _______
| | | |  |x| | |  | |x| |  | | |x|  |x|x| |  |x| |x|  | |x|x|  |x|x|x|
_______  _______  _______  _______  _______  _______  _______  _______
  |1|      |2|      |3|      |4|      |5|      |6|      |7|      |8|  
```

## Build
Clojure requires Java.  It's highly recommended to use Leiningen to run this Clojure project.
All commands are executed in this folder in a terminal.

## Usage
Start the program by opening a terminal window in this directory and then running

`lein run`

You will be prompted with the following options:
1. The rule to determine the initial state of the automaton.
1. The width of the grid.  A larger width displays more of the automaton's behavior.
1. The number of generations to calculate.  Default is equal to the width (enter a blank value).

## Examples
- Run the interactive terminal program:
`lein run`

- Run unit tests:
`lein test`

- Generate coverage from unit tests:
`lein cloverage`

### Interesting Rules
- 177
- 135

### To do
- [ ] Add input validation.

## License

Distributed under the GNU General Public License version 3.

See https://choosealicense.com/licenses/gpl-3.0/ or the included LICENSE file for full text.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]

## [0.1.0] - 2023-05-27
### Added
- Add terminal visualization of elementary cellular automata.

[Unreleased]: https://github.com/davis8dd/cellular-automata/compare/0.1.0...HEAD
[0.1.0]: https://github.com/davis8dd/cellular-automata/tree/0.1.0


test/cellular_automata/core_test.clj
(ns cellular-automata.core-test
  (:require [clojure.test :refer :all]
            [cellular-automata.core :refer :all]))

(deftest test-create-rule-with-int-less-than-0
  (testing "Validate boolean array generated by a rule."
    (is (= nil (create-rule -1)))))

(deftest test-create-rule-with-int-greater-than-255
  (testing "Validate boolean array generated by a rule."
    (is (= nil (create-rule 256)))))

(deftest test-create-rule-with-valid-int
  (testing "Validate boolean array generated by a rule."
    (let [expected-value {[true true true] false,
                          [true true false] true,
                          [true false true] true,
                          [true false false] false,
                          [false true true] true,
                          [false true false] true,
                          [false false true] true,
                          [false false false] false}]
      (is (= expected-value (create-rule 110))))))

(deftest test-pad-sides-with-given-char-expect-padded-sequence
  (testing "Test that padding a sequence with a character returns the expected padded sequence."
    (let [a-seq [1 2 3]
          padding-value \#
          expected-result [\# 1 2 3 \#]]
      (is (= (pad-sides-with a-seq padding-value) expected-result)))))

(deftest test-pad-sides-with-given-sequence-padding-expect-padding-with-sequence
  (testing "Test that padding a sequence with a sequence returns the expected padded sequence."
    (let [input-value [1 2 3]
          expected-result [0 0 1 2 3 0 0]]
      (is (= (pad-sides-with input-value [0 0]) expected-result)))))

(deftest test-int-to-bool-array-with-valid-8-bit-int-expect-bool-flag-array
  (testing "Passing a valid int generates a boolean array matching the integer as bits."
    (let [expected-value [true true true false true false false true]]
      (is (= (int-to-bool-array 233) expected-value)))))

(deftest test-int-to-bool-array-with-valid-6-bit-int-expect-8-bit-bool-flag-array
  (testing "Passing a valid int generates a boolean array matching the integer as bits."
    (let [expected-value [false false true false false false true false]]
      (is (= (int-to-bool-array 34) expected-value)))))

(deftest test-create-initial-row-given-row-width-expect-row-with-single-middle-value-set-to-true
  (testing "Test create a row of false boolean values, with the exception of the middle which is 'on'."
    (let [row-width 7
          expected-result [false false false true false false false]]
      (is (= (create-initial-row row-width) expected-result)))))

(deftest test-create-initial-row-with-even-length-expect-one-less-length-and-single-middle-value-set-to-true
  (testing "Test create an odd row of false boolean values, removing one from an even-length row if necessary."
    (let [row-width 8
          expected-result [false false false true false false false]]
      (is (= (create-initial-row row-width) expected-result)))))


(deftest test-pad-front-to-length-with-8-given-empty-and-space-expect-8-spaces
  (testing "Expect a thing.")
    (let [initial-array [1 2 3]
          padding 0
          final-size 5
          expected-value [0 0 1 2 3]]
      (is (= (pad-front-to-length-with initial-array final-size padding) expected-value))))

(deftest test-calculate-next-generation-given-valid-rule-expect-next-row-generated-from-rule
  (testing "That the passed rule is used to generate the next row."
    (let [a-vect (vector false false false true false false false)
          subvec-size 3
          rule {[false false false] true,
                [true false false] false,
                [false true false] false,
                [false false true] true,
                [true true false] true,
                [true false true] false,
                [false true true] true,
                [true true true] true}
          expected-result [true true false false true]]
      (is (= (calculate-next-generation a-vect subvec-size rule) expected-result)))))

(deftest test-calculate-grid-given-valid-rule-and-row-expect-calculated-grid
  (testing "That a valid rule and initial generation generates a grid."
    (let [valid-rule {[false false false] true,
                [true false false] false,
                [false true false] false,
                [false false true] true,
                [true true false] true,
                [true false true] false,
                [false true true] true,
                [true true true] true}
          number-of-rows 3
          initial-row [false false true false false]
          expected-result [[false false true false false]
                           [false true false false false]
                           [false false false true false]
                           [false true true false false]]]
      (is (= (calculate-grid valid-rule number-of-rows initial-row) expected-result)))))

(deftest test-str-to-int-given-int-in-str-expect-int-of-same-value
  (testing "Passing an integer as a string returns that integer."
    (let [input-value "123"
          expected-result 123]
      (is (= (str-to-int input-value) expected-result)))))

(deftest test-interactive-cellular-automata-given-valid-input-expect-no-errors
  (testing "That a valid rule and user input generates a grid."
    (with-in-str "20\n3\n3"
      (let [expected-result nil]
       (is (= (interactive-cellular-automata) expected-result))))))


test/cellular_automata/input_test.clj
(ns cellular-automata.input-test
  (:require [clojure.test :refer :all]
            [cellular-automata.input :refer :all]))

(deftest test-ask-input-given-input-expect-input-returned
  (testing "When providing a default value and input given, expect input returned."
    (with-in-str "testinput"
      (let [default-value "default-response"
            expected-value "testinput"]
        (is (= (ask-input default-value) expected-value))))))

(deftest test-ask-input-given-no-input-expect-default-returned
  (testing "When providing a default value and no input, expect default value returned."
    (with-in-str ""
      (let [input-value "default-response"]
        (is (= (ask-input input-value) input-value))))))

(deftest test-ask-input-with-message-given-response-expect-value-returned
  (testing "When prompted and input provided, expect input returned."
    (with-in-str "abc"
      (let [expected-value "abc"]
        (is (= (ask-input-with-message "amessage") expected-value))))))

(deftest test-ask-input-with-message-given-response-to-int-expect-int-value-returned
  (testing "When prompted and input transformed via function, expect transformed value returned."
    (with-in-str "123"
      (let [expected-value 123]
        (is (= (ask-input-with-message "amessage" #(Integer. %)) expected-value))))))

(deftest test-ask-input-with-message-and-default-given-no-response-expect-default-as-int-value-returned
  (testing "When prompted and default value transformed via function, expect transformed value returned."
    (with-in-str ""
      (let [default-value "234"
            expected-value 234]
        (is (= (ask-input-with-message "amessage" #(Integer. %) default-value) expected-value))))))


test/cellular_automata/graphics_test.clj
(ns cellular-automata.graphics-test
  (:require [clojure.test :refer :all]
            [cellular-automata.graphics :refer :all]))

(deftest test-string-repeating-char-given-character-and-times-to-repeat-expect-character-repeated-that-many-times
  (testing "Test that the given character is repeated in a string the specified number of times."
    (let [a-char \a
          times-to-repeat 5
          expected-result "aaaaa"]
      (is (= (string-repeating-char a-char times-to-repeat) expected-result)))))

(deftest test-row-to-render-row-given-boolean-sequence-expect-symbol-sequence-returned
  (testing "asdf"
    (let [input-value [true false true]
          expected-result [:on :off :on]]
      (is (= (row-to-render-row input-value) expected-result)))))

(deftest test-render-row-to-string-given-????-expect???
  (testing "Render array of characters given character keys as input."
    (let [input-value [:on :off :on]
          expected-result (str (get-in render-chars [:cells :on]) (get-in render-chars [:cells :off]) (get-in render-chars [:cells :on]))]
      (is (= (render-row-to-string input-value) expected-result)))))

(deftest test-grid-to-render-of-single-character-given-border-expect-3x3-grid-with-border
  (testing "Render single character with a border, returning a 3x3 grid."
    (let [character-to-render \x
          grid-to-render [[character-to-render character-to-render] [character-to-render character-to-render] [character-to-render character-to-render]]
          expected-value [[(get-in render-chars [:border :top-left]) (get-in render-chars [:border :top]) (get-in render-chars [:border :top]) (get-in render-chars [:border :top-right])]
                          [(get-in render-chars [:border :left]) character-to-render character-to-render (get-in render-chars [:border :right])]
                          [(get-in render-chars [:border :left]) character-to-render character-to-render (get-in render-chars [:border :right])]
                          [(get-in render-chars [:border :left]) character-to-render character-to-render (get-in render-chars [:border :right])]
                          [(get-in render-chars [:border :bottom-left]) (get-in render-chars [:border :bottom]) (get-in render-chars [:border :bottom]) (get-in render-chars [:border :bottom-right])]]]
      (is (= expected-value (with-border render-chars grid-to-render))))))


src/cellular_automata/core.clj
(ns cellular-automata.core
    (:gen-class))

(require '[cellular-automata.graphics :as graphics]
         '[cellular-automata.input :as input])

(def rule-keys [[true true true]
                [true true false]
                [true false true]
                [true false false]
                [false true true]
                [false true false]
                [false false true]
                [false false false]])

(def ui-messages {:greeting "Welcome to the elementary cellular automata generator!"
                  :grid-width "Please enter a grid width."
                  :generations "How many generations would you like to see?"
                  :initial-row "Do you want to set the initial generation?"
                  :which-rule "Which rule would you like to use (Enter a number between [0, 255], or h for help)"
                  :exiting "Goodbye!"})

(defn pad-front-to-length-with
    "Add a number of padding values to the front of a sequence to make it the specified length."
    [a-coll final-length padding-value]
    (let [padding-size (- final-length (count a-coll))]
    (into [] (flatten
                 (concat (repeat padding-size padding-value) [a-coll])))))

(defn int-to-bool-array
    "Change integer to a boolean array to represent cell on/off states."
    [an-int]
    (pad-front-to-length-with (do (map #(= 1 %)
                                       (map #(Integer. (str %))
                                            (Integer/toBinaryString an-int)))) 8 false))

(defn create-rule
    "Create rule for processing next generation"
    [an-int]
    (cond
      (< an-int 0) nil
      (> an-int 255) nil
      :else (zipmap rule-keys (int-to-bool-array an-int)))) ;; Return a set of (front-value, on/off) pairs for all front-things

(defn convert-str-to-type
    [constructor-fn input-string]
    (try (constructor-fn input-string)
         (catch NumberFormatException e (println "Please enter an integer."))))

(def str-to-int
    "Convert a string to an Integer, else throw an exception"
    (partial convert-str-to-type #(Integer. %)))

(defn pad-sides-with
    "Add supplied value to front and back of a sequence."
    [a-coll padding-value]
    (into [] (flatten
                 (concat [padding-value] [a-coll] [padding-value]))))

(defn create-initial-row
    "Create a starting row of odd width with one 'on' cell in center of row"
    [width]
    (let [row-width (str-to-int width)]
          (pad-sides-with true (repeat (- (/ row-width 2) (rem (+ width 1) 2)) false))))

(defn calculate-next-generation
    "Process a generation using a rule and return next generation.
     The rule specifies each element in the next generation
     based on the current generation."
    [a-row subvec-size rule]
    (let [row-length (count a-row)]
    (loop [next-row [] counter 0]
        (if (> counter (- row-length subvec-size))
            next-row
            (recur (into next-row (vector (get rule (subvec a-row counter (+ subvec-size counter))))) (inc counter))))))

(defn get-next-row
    "Return next row in grid by calculating new generation."
    [a-vect rule]
    (pad-sides-with
        (calculate-next-generation
            a-vect
            3
            rule)
        false))

(defn calculate-grid
    "Calculate grid of specified length using either a supplied or default starting generation."
    ([rule num-of-rows initial-row]
        (loop [grid [initial-row] counter 1]
            (if (> counter num-of-rows)
                grid
                (recur (conj grid (into (vector) (get-next-row (last grid) rule))) (inc counter)))))
    ([rule num-of-rows]
        (calculate-grid rule num-of-rows [false false false false true false false false false])))

(defn interactive-cellular-automata
    "Prompt user for input and display corresponding cellular automoata"
    []
    (do (println (:greeting ui-messages))
        (let [rule (input/ask-input-with-message (:which-rule ui-messages) str-to-int)

              grid-width (input/ask-input-with-message (:grid-width ui-messages) str-to-int)
              grid-height (input/ask-input-with-message (:generations ui-messages) str-to-int)
              the-grid ()]
             (graphics/display-grid (graphics/with-border graphics/render-chars (graphics/rendered-grid (calculate-grid (create-rule rule) grid-height (create-initial-row grid-width))))))
        (println (:exiting ui-messages))))

(defn -main
    "Run interactive cellular automata function"
    [& args]
    (interactive-cellular-automata))


src/cellular_automata/input.clj
(ns cellular-automata.input
    (:gen-class))

 (defn ask-input
     "Ask for input, optionally returning a default value.  May default to nil, indicating no input and no default value."
     ([]
         (ask-input nil))
     ([default-value]
         (let [input-value (read-line)]
             (if (empty? input-value)
                 default-value
                 input-value))))

 (defn ask-input-with-message
     "Prompt user for input with message, optionally display default value."
     ([message]
         (do (println message)
             (ask-input)))
     ([message converter-fn]
         (do (println (str message  ":"))
             (converter-fn (ask-input))))
     ([message converter-fn default-value]
         (do (println (str message  ": [" default-value "]"))
             (converter-fn (ask-input default-value)))))


src/cellular_automata/graphics.clj
(ns cellular-automata.graphics
    (:gen-class))

(def render-chars {:cells {:on (char 9607)
                           :off (char 9617)}
                           :border {:top (char 9552)
                           :bottom (char 9552)
                           :left (char 9553)
                           :right (char 9553)
                           :top-right (char 9559)
                           :bottom-right (char 9565)
                           :top-left (char 9556)
                           :bottom-left (char 9562)}})

(defn display-grid
    "Display rendered grid of cellular automata."
    [rendered-grid]
    (doseq [rendered-grid-row rendered-grid]
        (println (apply str rendered-grid-row))))

(defn string-repeating-char
    "Repeat supplied character"
    [a-char times]
    (apply str (repeat times a-char)))

(defn add-border-top
    "Add top border"
    [render-view-chars border-width]
        (let [top-left-char (get-in render-view-chars [:border :top-left])
              top-right-char (get-in render-view-chars [:border :top-right])
              top-char (get-in render-view-chars [:border :top])]
                (concat (vector top-left-char)
                 (repeat border-width top-char)
                 (vector top-right-char))))

(defn add-border-sides-to-row
    "Add border sides to single row"
    [render-view-chars render-row]
        (let [right-char (get-in render-view-chars [:border :right])
              left-char (get-in render-view-chars [:border :left])]
                  (concat (vector left-char) render-row (vector right-char))))

(defn add-border-sides
    "Add border sides to group of rows"
    [render-view-chars render-grid]
        (let [right-char (get-in render-view-chars [:border :right])
              left-char (get-in render-view-chars [:border :left])]
                 (reduce (fn [bordered-grid render-row] (conj bordered-grid (add-border-sides-to-row render-chars render-row))) []  render-grid)))

(defn add-border-bottom
    "Add border bottom"
    [render-view-chars border-width]
        (let [bottom-left-char (get-in render-view-chars [:border :bottom-left])
              bottom-right-char (get-in render-view-chars [:border :bottom-right])
              bottom-char (get-in render-view-chars [:border :bottom])]
                (concat (vector bottom-left-char)
                 (repeat border-width bottom-char)
                 (vector bottom-right-char))))


(defn with-border
    "Display border of characters defined by 'render-chars' around content"
    [render-chars content]
        (let [content-width (count (first content))]
                (-> [(add-border-top render-chars content-width)]
                    (into (add-border-sides render-chars content))
                    (into [(add-border-bottom render-chars content-width)]))))

(defn row-to-render-row
    "Take a generation and return a list consumable by the View functions."
    [row]
    (reduce (fn [render-row cell] (if cell (conj render-row :on) (conj render-row :off))) [] row))

(defn render-row-to-string
    "Take a View-friendly generation and return a string ready for display rendering."
    [render-row]
    (apply str (map #(get (get render-chars :cells) %) render-row)))

(defn rendered-grid
    [grid] (->> grid
                (map row-to-render-row)
                (map render-row-to-string)))


src/cellular_automata/output/terminal.clj
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Render functions
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; THIS HANDLES ALL SIDE-EFFECT FUNCTIONS

;; Separated rendering functions only for terminal... this should only handle strings
;; that are to be directly rendered to the terminal.  Higher-level business logic etc
;; should be in a separate namespace that interfaces model and view

(ns cellular-automata.output)
;; An example of displaying a grid.  This should probably be a unit test
;; This one looks neat!
;(display-grid (calculate-grid (create-rule 177) 50 [false false false false false false false false false false true false false false false false false false false false false]))

(defn display-grid
  "Display rendered grid of cellular automata"
  [rendered-grid]
  (doseq [rendered-grid-row (vec rendered-grid)] (println rendered-grid-row)))


(def render-chars {:cells {:on (char 9607)
;                              :off \ }
                              :off (char 9617)}
                   :border {:top (char 9552)
                               :bottom (char 9552)
                               :left (char 9553)
                               :right (char 9553)
                               :top-right (char 9559)
                               :bottom-right (char 9565)
                               :top-left (char 9556)
                               :bottom-left (char 9562)}})


;;(defn print-chars [chars-rendered] (doseq [a-char chars-rendered] (println (str "*** " a-char " ***"))))

;; Abstractly, display a grid.  This should eventually call an imported display library based on the output device used
(defn print-test-border [render-chars] (let [tl ((comp :top-left :border) render-chars)
                                                t ((comp :top :border) render-chars)
                                                tr ((comp :top-right :border) render-chars)
                                                l ((comp :left :border) render-chars)
                                                r ((comp :right :border) render-chars)
                                                bl (get-in render-chars [:border :bottom-left])
                                                b (get-in render-chars [:border :bottom])
                                                br (get-in render-chars [:border :bottom-right])
                                                on (get-in render-chars [:cells :on])
                                                off (get-in render-chars [:cells :off])]
                                          (println (str tl t t t tr))
                                          (println (str l off off off r))
                                          (println (str l off on off r))
                                          (println (str l off off off r))
                                          (println (str bl b b b br))))


(defn string-repeating-char [a-char times] (apply str (repeat times a-char)))
(defn add-border-top [render-view-chars border-width] (let [top-left-char (get-in render-view-chars [:border :top-left])
                                    top-right-char (get-in render-view-chars [:border :top-right])
                                    top-char (get-in render-view-chars [:border :top])]
                              (println "Border has width of " border-width) (apply str [top-left-char (string-repeating-char top-char border-width) top-right-char])))

(defn add-border-sides-to-row [render-view-chars render-row] (let [right-char (get-in render-view-chars [:border :right])
                                    left-char (get-in render-view-chars [:border :left])] (apply str (concat (str left-char) render-row (str right-char)))))

(defn add-border-sides [render-view-chars render-grid] (let [right-char (get-in render-view-chars [:border :right])
                                    left-char (get-in render-view-chars [:border :left])]
                              (reduce (fn [bordered-grid render-row] (conj bordered-grid (add-border-sides-to-row render-chars render-row))) []  render-grid)))
(defn add-border-bottom [render-view-chars border-width] (let [bottom-left-char (get-in render-view-chars [:border :bottom-left])
                                    bottom-right-char (get-in render-view-chars [:border :bottom-right])
                                    bottom-char (get-in render-view-chars [:border :bottom])]
                              (println "Width has value of " border-width) (apply str [bottom-left-char (string-repeating-char bottom-char border-width) bottom-right-char])))

;;(defn print-with-border [grid] (println "Grid has value of " grid) (add-border-sides grid) (add-border-top (count (first grid))) (add-border-bottom (count (first grid))))


;; SOMETHING LIKE THIS:
;; (->> a-grid (concat [["a" "b" "c"]]) (concat [["x" "y" "z"]])) 
;;(defn add-border [grid] (->> (add-border-sides grid) (concat [(add-border-top (count (first grid)))]) (concat [(add-border-bottom (count (first grid)))]) (map println) ))
(defn add-border [render-chars grid] (doseq [grid-row (vector (add-border-top render-chars (count (first grid))) (add-border-sides render-chars grid) (add-border-bottom render-chars (count (first grid))))] (println grid-row)))

(defn with-border [render-chars content]
    (let [content-width (count (first content))] (-> [(add-border-top render-chars content-width)]
                                                     (into (add-border-sides render-chars content))
                                                     (into [(add-border-bottom render-chars content-width)]))))

;; for each row in grid, add border-left to front and border-right to end of row
;; add new row to front of grid that's border top
;; add new row to back of grid that's border bottom

(defn render-row-to-str [render-row] (apply str (map #(get (get render-chars :cells) %) render-row)))
(defn row-to-render-row [row] (reduce (fn [render-row cell] (if cell (conj render-row :on) (conj render-row :off))) [] row))
;(defn rendered-grid [grid] (map render-row-to-str (map row-to-render-row grid)))
(defn rendered-grid [grid] (->> grid (map row-to-render-row) (map render-row-to-str)))
;; (map println (map render-row-to-str (map row-to-render-row (calculate-grid (create-rule 172) 20 [false true false false true true false true false]))) )

;;;;;;;;;;;;;;;;;;
;; full rendering:
; USE CASE 1: no border
(display-grid (rendered-grid (calculate-grid (create-rule 177) 45 [false false false false false false false false false false true false true false false false false false false false false])))
;fail;
; USE CASE 2: with border
(display-grid (with-border render-chars (rendered-grid (calculate-grid (create-rule 177) 45 [false false false false false false false false false false true false true false false false false false false false false]))))
;(display-grid (add-border render-chars (rendered-grid (calculate-grid (create-rule 177) 45 [false false false false false false false false false false true false true false false false false false false false false]))))
;
;;;;;;;;;;;;;;;;;;

;;  (map println (map render-row-to-str (map row-to-render-row rendered-grid))))

